src/graph/immutable_graph.cc

```diff
@@ -0,0 +1,975 @@
+  const auto dstlen = dst_ids->shape[0];
+  int64_t i, j;
+
+  CHECK((srclen == dstlen) || (srclen == 1) || (dstlen == 1))
+    << "Invalid src and dst id array.";
+
+  const int src_stride = (srclen == 1 && dstlen != 1) ? 0 : 1;
+  const int dst_stride = (dstlen == 1 && srclen != 1) ? 0 : 1;
+  const dgl_id_t* src_data = static_cast<dgl_id_t*>(src_ids->data);
+  const dgl_id_t* dst_data = static_cast<dgl_id_t*>(dst_ids->data);
+
+  std::vector<dgl_id_t> src, dst, eid;
+
+  for (i = 0, j = 0; i < srclen && j < dstlen; i += src_stride, j += dst_stride) {
+    const dgl_id_t src_id = src_data[i], dst_id = dst_data[j];
+    CHECK(HasVertex(src_id) && HasVertex(dst_id)) <<
+        "invalid edge: " << src_id << " -> " << dst_id;
+
+    std::pair<const dgl_id_t *, const dgl_id_t *> edges;
+    if (this->in_csr_)
+      edges = this->GetInEdgeIdRef(src_id, dst_id);
+    else
+      edges = this->GetOutEdgeIdRef(src_id, dst_id);
+
+    size_t len = edges.second - edges.first;
+    for (size_t i = 0; i < len; i++) {
+        src.push_back(src_id);
+        dst.push_back(dst_id);
+        eid.push_back(edges.first[i]);
+    }
+  }
+
+  int64_t rstlen = src.size();
+  IdArray rst_src = IdArray::Empty({rstlen}, src_ids->dtype, src_ids->ctx);
+  IdArray rst_dst = IdArray::Empty({rstlen}, src_ids->dtype, src_ids->ctx);
+  IdArray rst_eid = IdArray::Empty({rstlen}, src_ids->dtype, src_ids->ctx);
+  dgl_id_t* rst_src_data = static_cast<dgl_id_t*>(rst_src->data);
+  dgl_id_t* rst_dst_data = static_cast<dgl_id_t*>(rst_dst->data);
+  dgl_id_t* rst_eid_data = static_cast<dgl_id_t*>(rst_eid->data);
+
+  std::copy(src.begin(), src.end(), rst_src_data);
+  std::copy(dst.begin(), dst.end(), rst_dst_data);
+  std::copy(eid.begin(), eid.end(), rst_eid_data);
+
+  return ImmutableGraph::EdgeArray{rst_src, rst_dst, rst_eid};
+}
+
+ImmutableGraph::EdgeArray ImmutableGraph::Edges(bool sorted) const {
+  int64_t rstlen = NumEdges();
+  IdArray rst_src = IdArray::Empty({rstlen}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  IdArray rst_dst = IdArray::Empty({rstlen}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  IdArray rst_eid = IdArray::Empty({rstlen}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
+  dgl_id_t* rst_src_data = static_cast<dgl_id_t*>(rst_src->data);
+  dgl_id_t* rst_dst_data = static_cast<dgl_id_t*>(rst_dst->data);
+  dgl_id_t* rst_eid_data = static_cast<dgl_id_t*>(rst_eid->data);
+
+  auto out_csr = GetOutCSR();
+  // If sorted, the returned edges are sorted by the source Id and dest Id.
+  for (size_t i = 0; i < out_csr->indptr.size() - 1; i++) {
+    std::fill(rst_src_data + out_csr->indptr[i], rst_src_data + out_csr->indptr[i + 1],
+              static_cast<dgl_id_t>(i));
+  }
+  std::copy(out_csr->indices.begin(), out_csr->indices.end(), rst_dst_data);
+  std::copy(out_csr->edge_ids.begin(), out_csr->edge_ids.end(), rst_eid_data);
+
+  // TODO(zhengda) do I need to sort the edges if sorted = false?
```

Comment:
It should be in the edge id order if sorted=False

Reply:
i see the C++ documentation says that. However, sort all edges in their eid order can be very expensive in immutable graph. It seems this is an common operations. Is this sementics used anywhere?